### PR TITLE
FB0D and methane iron OC properties updates

### PR DESCRIPTION
**Pull request by bpaul4**
## Fixes
- Add unit consistency checks to FB0D and methane iron OC test modules
- Updates to FB0D model addressing unit inconsistency issues and derived units
- Updated units and scaling in methane iron OC properties

## Summary/Motivation:
- Updating FB0D and associated files to new IDAES units standards

## Changes proposed in this PR:
-
-

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by bpaul4** [comment]
Tests may fail due to recent unrelated updates to the test_commands script. All gas_solid_contactors tests pass when run locally.

**Review by Robbybp — COMMENTED** [review]
Very happy to see these changes implemented. Have some clarifying questions.

File: idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py
Comment (bpaul4) [reply idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py:476]: I see that the oxidation gas thermo package had this updated in #416, I can update this here as well.

File: idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py
Comment (bpaul4) [reply idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py:541]: Yes, the 1/Kelvin units are to address unit inconsistency issues.

File: idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/solid_phase_thermo.py
Comment (bpaul4) [reply idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/solid_phase_thermo.py:605]: Some of the component mole fractions may become very small, either through inlet specifications (see the MB flowsheet example where the inlet water fraction is 1E-5) or if methane approaches full conversion by the reactor outlet. The scaling helps address convergence issues; @chineduobiora may be able to expand upon the reasoning behind the scaling factor.

File: idaes/gas_solid_contactors/unit_models/fixed_bed_0D.py
Comment (bpaul4) [reply idaes/gas_solid_contactors/unit_models/fixed_bed_0D.py:222]: Pyomo does not currently support units for ContinuousSet objects, and @andrewlee94 has opened an issue requesting this feature. Since we cannot specify mass/time units here but the numerical calculations are still correct, the accumulation variables are manually corrected for units in the accumulation constraints.

File: idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/solid_phase_thermo.py
Comment (andrewlee94) [reply idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/solid_phase_thermo.py:605]: Assuming this is a sum of mole fractions, you should try to get the currently defined scaling factor first, with a fall back to a default if one has not been defined.

File: idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py
Comment (andrewlee94) [reply idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py:541]: I think you need to be a bit more careful here. The first reference to temperature needs to have the units removed as it is raised to a variable, non-integer power. However, in the later terms the parameters should have units which cancel with T - e.g. parameter 3 should have units of `1/K` which will cancel with those of `temperature` as required. Similar for parameter 4 (units should be 1/K**2).

Note that this required you to create separate `Vars` for each parameter (rather than a single indexed parameter) as all elements of an indexed `Var` must have the same units. However, this is what we are doing everywhere else in IDAES, and is probably the better way to do this.

File: idaes/gas_solid_contactors/unit_models/fixed_bed_0D.py
Comment (andrewlee94) [reply idaes/gas_solid_contactors/unit_models/fixed_bed_0D.py:222]: @bpaul4 I am not sure that this is correct. In the 1D control volume, we assign the correct units to the `DerivativeVar` (see line 424 in CV1D). I am not sure how we made this work, but it does.

Comment (bpaul4) [reply idaes/gas_solid_contactors/unit_models/fixed_bed_0D.py:222]: It appears line 268 in the 0D control volume does something similar, although Pyomo threw an InconsistentUnits error when I attempted to set expected units on the derivative variable itself. An internal constraint defined the discretization as accumulation = Δ(mass)/Δ(time) with units [m/s] = [m]/[dimensionless since ContinuousSets are not passed with units].

File: idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py
Comment (bpaul4) [reply idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py:541]: I'm reviewing property packages with Parameters for each coefficient, indexed by component, for proper syntax. Would you recommend modifying the gas/solid Cp and gas thermo_cond parameters in this way as well?

Comment (andrewlee94) [reply idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py:541]: It needs to happen at some point. However, if the PR starts getting too big you might want to try breaking it up into stages.

Comment (bpaul4) [reply idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py:541]: This PR is focusing only on the FixedBed0D model and related updates  to the methane iron OC properties. There will be other PRs for the MovingBed and BubblingFluidizedBed models, and I could leave the Param updates until then to simplify this PR.

File: idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/solid_phase_thermo.py
Comment (Robbybp) [reply idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/solid_phase_thermo.py]: I have no real problem with this scaling factor, but my intuition is that adding a scaling factor of 1e2 to a constraint actually makes it _harder_ to converge. I.e., we need to converge our mole fraction variables to a tighter tolerance to satisfy this constraint to an absolute tolerance of 1e-8. This factor also increases the constraint's Jacobian entries from 1 to 100, which we may or may not want? 

Again no real problem here, I'm just being difficult because I'm trying to arrive at a more systematic understanding of scaling myself and am always curious about hard-coded scaling factors.

File: idaes/gas_solid_contactors/unit_models/fixed_bed_0D.py
Comment (Robbybp) [reply idaes/gas_solid_contactors/unit_models/fixed_bed_0D.py:222]: Okay, I think I see. The problem is that the "Delta t" constants in the time discretization equation don't have units. It's not a problem for the space discretization equations because our length domains are normalized (dimensionless).

File: idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py
Comment (chineduobiora) [inline idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py]: The example gas proppack we went through during our last meeting has an updated code for the cp param. If you get a chance, please take a look at it? I think it is preferable to the setup shown here as the parameters are grouped by units. The same idea is applicable to the cp parameters in the solids propack, and to the parameters for viscosity, thermal conductivity, and diffusivity in the gas proppack.

Comment (bpaul4) [reply idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py]: @chineduobiora the most updated version from commit "Replacing dimensionless parameter dictionary with dimensional parameters" has the updated setup - the coefficients are extracted to 8 parameters, each indexed by the component list.

File: idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py
Comment (chineduobiora) [reply idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py]: I think that's the approach to take. @bpaul4 the gas proppack file we discussed at our last meeting has the cp param, and enthalpy and cp_mol variables setup in this manner. Please take a look when you get the opportunity.

File: idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py
Comment (chineduobiora) [inline idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py:215]: This unit is kg * m**-1 * s**-1 * K**-visc_d_param_2

Comment (chineduobiora) [reply idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py:215]: The equation is in this form: k=(a*(T^b))/(1+c/T +d/(T^2 ))

File: idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py
Comment (chineduobiora) [inline idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py:252]: This unit is J * m**-1 * s**-1 * K**-[1 + therm_cond_param_2]

File: idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py
Comment (bpaul4) [reply idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py:215]: Yes, this form is used for thermal conductivity and viscosity. If the units of one of those variables is [x], then the parameters should have the units {a: [x], b:[], c: [K], d: [K^2]}, correct? In the expressions the first T is divided by [K] before the exponent is applied.

File: idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py
Comment (chineduobiora) [inline idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py:628]: Division by pyunits.K will no longer be necessary when the units for visc_d_param_1 are updated per prior comment.

File: idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py
Comment (chineduobiora) [inline idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py:701]: Division by pyunits.K will no longer be necessary when the units for therm_cond_param_1 are updated per prior comment.

File: idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/solid_phase_thermo.py
Comment (chineduobiora) [reply idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/solid_phase_thermo.py]: 1e2 was originally hardcoded here because my experience is to have the mole fraction variables scaled by at least a factor of 100. For the generic application I would lean towards @andrewlee94's suggestion to scale this based on a default scaling factor (using the default scaling for the mole fraction variables is a good choice). However, to @Robbybp's point the "right" scaling factor also tends to be problem specific, and a user should be able to adjust scaling factors from the flowsheet level as needed.

Comment (chineduobiora) [reply idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/solid_phase_thermo.py]: @bpaul4 the gas proppack we discussed at our last meeting has examples of how default scaling can be set for individual variables (see the end of the PhysicalParameterData Class), and how they can be applied to scale constraints.

File: idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py
Comment (chineduobiora) [inline idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py:900]: What does _k_rxn represent?

File: idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py
Comment (chineduobiora) [inline idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py:904]: What does _k_rxn represent?

Comment (bpaul4) [reply idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py:904]: That is a typing error, I will remove the extra attributes on those lines.

File: idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py
Comment (chineduobiora) [reply idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py:215]: "In the expressions the first T is divided by [K] before the exponent is applied." That works, but it might get confusing for someone who compares the code to the actual equation.

File: idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py
Comment (chineduobiora) [inline idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py:945]: This is only one equation (`1 = sum(x_i)`), so the `for loop` isn't required. The component index should also be specified i.e. choose one of the components from the list.

File: idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py
Comment (chineduobiora) [inline idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py:953]: Update these in the same way as the scaling for the sum_component_eqn constraint i.e. define default scaling for visc_d and therm_cond like you did for mole_frac_comp, and then use them to set scaling factors for the constraints.

File: idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/hetero_reactions.py
Comment (chineduobiora) [inline idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/hetero_reactions.py:170]: For consistency purposes, please change all `N_reaction` occurences to `rxn_order`.

File: idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/hetero_reactions.py
Comment (chineduobiora) [inline idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/hetero_reactions.py:406]: `OC_conv` and `OC_conv_temp` have dimensionless units.

File: idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/solid_phase_thermo.py
Comment (chineduobiora) [inline idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/solid_phase_thermo.py:52]: The suggested changes to the gas property package also apply to the solid property package.

**Review by chineduobiora — COMMENTED** [review]
Happy New Year @bpaul4. The changes so far look good and I think we are close to completion of this PR, but I have included a couple more minor comments to be addressed.

File: idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py
Comment (bpaul4) [reply idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py:953]: @chineduobiora Would you recommend updating the reaction module scaling in this manner as well?

Comment (chineduobiora) [reply idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/gas_phase_thermo.py:953]: Yes, that will be good. However, unlike the physical properties blocks you can't set scaling factors for variables using the set_default_scaling method in the reaction properties block (it doesn't seem to be supported). So instead, I suggest using the set_scaling_factor method to set the scaling factors for variables under the calculate_scaling_factors method, and then you can set the scaling for the related constraints as before. See the example below:

```
    def calculate_scaling_factors(self):
        super().calculate_scaling_factors()

        def _set_default_factor(v, s):
            for i in v:
                if iscale.get_scaling_factor(v[i]) is None:
                    iscale.set_scaling_factor(v[i], s)
        _set_default_factor(self.reaction_rate, 1)
        _set_default_factor(self.rate_constant, 1e2)
```

File: idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/solid_phase_thermo.py
Comment (chineduobiora) [inline idaes/gas_solid_contactors/properties/methane_iron_OC_reduction/solid_phase_thermo.py:674]: Select a component from the solids component list. CH4 is only in the gas property package.

**Review by chineduobiora — DISMISSED** [review]
Looks good to me.

**Review by andrewlee94 — APPROVED** [review]
I don't see anything wrong, and I'll trust @chineduobiora for the details.

**Comment by andrewlee94** [comment]
There are some issues that showed up in pyling that need to be fixed - in lines 968 and 973 of the gas phase thermo properties you are missing an argument to  `iscale.constraint_scaling_transform` (the constraint to be transformed). This also suggests that you are not testing the scaling methods completely - you should have a test the runs the scaling method with all properties present and make sure that the correct scaling factor is set for each.

The other failing test is a known issue separate from this PR. It will be fixed as part of #647.

**Comment by chineduobiora** [comment]
> There are some issues that showed up in pyling that need to be fixed - in lines 968 and 973 of the gas phase thermo properties you are missing an argument to `iscale.constraint_scaling_transform` (the constraint to be transformed). This also suggests that you are not testing the scaling methods completely - you should have a test the runs the scaling method with all properties present and make sure that the correct scaling factor is set for each.
> 
> The other failing test is a known issue separate from this PR. It will be fixed as part of #647.

Mistake was from me. I've made the update. Yes, we probably should have a test for the scaling method.

**Review by chineduobiora — APPROVED** [review]

**Review by andrewlee94 — APPROVED** [review]
Looks good.

**Comment by Robbybp** [comment]
I have a fair amount of research code that depends on these property packages, and some of it is failing when I use this branch, so I'd like to figure that out before this PR is merged. Most of it is probably that I need to adjust ranges on some of my tests, so hopefully it won't take too long to sort through.

**Comment by Robbybp** [comment]
Okay, most of the errors I encountered I could fix by changing a tolerance or something, but I have one dynamic optimization problem that solves with the old version of the property packages (with hard-coded scaling factors on sum_component_eqn, OC_conv_eqn, etc. constraints), but not the new version (without the hard-coded scaling factors). I'm not sure this is a compelling reason to re-introduce the factors, however, unless others report problems as well. Hopefully this is not too difficult for me to work around. Another option is that we could manually scale constraints with a unit conversion rather than hard-coded constants. Let me know your thoughts:
1. Has anybody done an convergence reliability studies on the old/new forms of the model?
2. Should we keep the factors on these constraints so as to not potentially break users' optimization problems?
3. If yes, should we replace the constant factors with unit conversions?

**Comment by chineduobiora** [comment]
@ My inclination is to minimize the number of variable and constraint scaling factors that are introduced to the model. I think it is best that a user updates the scaling of the variables and constraints as they deem fit, because the appropriate scale factors will vary by application (even for the same proppack). At the flowsheet level you can also update the scaling factors of individual constraints and variables to factors of your choice. A user can also update a local copy of their proppack with the scale factor of interest for a variable or constraint.

**Comment by chineduobiora** [comment]
@Robbybp, for the other questions, see below:

1. Has anybody done any convergence reliability studies on the old/new forms of the model?
     To the best of my knowledge we haven’t done this. We will consider this when updating the MB and BFB to have units.
2. Should we keep the factors on these constraints so as to not potentially break users' optimization problems?
     I’ll prefer not to keep the factors on the constraints. Users can update the factors in their local copies of the proppack, or at the flowsheet level as their application dicatates. The scaling module supports two methods (`extreme_jacobian_entries(m)`, and `badly_scaled_var_generator(m)`) that can allow a user investigate their model and update scaling factors as needed. If you aren’t already doing this, you’ll need to apply the scaling transform using `calculate_scaling_factors(m)` at the flowsheet level.
3. If yes, should we replace the constant factors with unit conversions?
    I recommend updating the constant factors with unit conversions (using the scaling tool).

**Comment by andrewlee94** [comment]
@Robbybp @chineduobiora Are we ready to merge this PR (assuming the tests pass)?

**Comment by chineduobiora** [comment]
@andrewlee94 I am happy for it to be merged. @Robbybp what are your thoughts? Are my responses to your comments on convergence reliability studies, and scaling sufficient?

**Comment by Robbybp** [comment]
@chineduobiora @andrewlee94 I'd prefer to not alter default algebraic form of models unless we have a good reason for doing so. A good reason might be "make dens_mass a variable rather than an expression so that it can be fixed, as it is a common degree of freedom." Another good reason would be "change a hard-coded factor on a constraint to make convergence more reliable for a suite of cases we are interested in." I'm not sure "make constraint expressions simpler" is a good enough reason to change the form of a model, given that doing this may break users' optimization problems.

Here I am being somewhat selfish -- I have an NMPC problem that I'm solving for a presentation that breaks when we change these constraint factors. I'm sure I can find a way to work around this change, but I imagine it would be a much more frustrating experience for a user if they update versions of IDAES and all of a sudden one of their optimization problems stops converging.

**Comment by chineduobiora** [comment]
@Robbybp The default algebraic forms of the models we currently have are not being changed in these PRs. The only changes being made are to add units (including using pyconvert to ensure that units are consistent in equations), and to remove hard-coded factors from constraints. The variable and constraint scaling transformations are optional and only work on the model when called. Users can update the scaling transformations as they deem fit for their particular problem (i.e. scale a particular constraint as needed to improve convergence). Also, we haven't applied these changes to the MB and BFB yet so there might be some inconsistencies at the moment if you using the proppacks in this PR with those models. The MB is next on @bpaul4 list, before the BFB.

Does this address your comment? If not we can discuss further at the CLC meeting on Thursday.

**Comment by Robbybp** [comment]
> and to remove hard-coded factors from constraints

This is what I mean by "changing the algebraic form." This can affect convergence, so I'd rather only do this if we have a good enough reason. Making the expressions simpler may be a good enough reason, but I'd like to bring this up at a core dev meeting to see what others think. On one hand, it seems very restrictive to say "we will never change a constraint expression ever," but on the other, it would be nice if model-solver performance/convergence changed as little as possible (ideally not at all) from one IDAES version to the next.

**Pull request merged**
